### ISSUE TITLE
Gesture Recognizer unable to use GPU inference

### ISSUE THREAD

**Issue by weizenyang** [google/mediapipe#4327]
### Have I written custom code (as opposed to using a stock example script provided in MediaPipe)

Yes

### OS Platform and Distribution

Windows 11

### MediaPipe version

Latest

### Solution

tasks-vision

### Programming Language and version

Javascript

### Describe the actual behavior

I Tried enabling GPU Inference in baseOption but got an error,
```
caught Error: You must specify a canvas for GPU processing.

```
```
			const gestureRecognizer = await GestureRecognizer.createFromOptions(vision, {
				baseOptions: {
					modelAssetPath: "../gesture_recognizer_pruned_new.task",
					delegate: 'GPU'				
				},
				runningMode: "video" 
			});
```

I then added a canvas to createFromOptions() method: 
```
			const gestureRecognizer = await GestureRecognizer.createFromOptions(vision, {
				baseOptions: {
					modelAssetPath: "../gesture_recognizer_pruned_new.task",
					delegate: 'GPU'				
				},
				runningMode: "video",
				canvas: canvas
			});
```
and got this error instead
```
caught Error: StartGraph failed: $Service "kGpuService", required by node mediapipe_tasks_vision_gesture_recognizer_gesturerecognizergraph__mediapipe_tasks_vision_hand_landmarker_handlandmarkergraph__mediapipe_tasks_vision_hand_detector_handdetectorgraph__mediapipe_tasks_components_processors_imagepreprocessinggraph__ImageCloneCalculator, was not provided and cannot be created: emscripten_webgl_create_context() returned error 0; StartRun failed
    at R.handleErrors (vision_bundle.js:1:438974)
    at R.setGraph (vision_bundle.js:1:438658)
    at R.refreshGraph (vision_bundle.js:1:651769)
    at vision_bundle.js:1:438078
    at async createTaskRunner (vision_bundle.js:1:436884)
    at async index.astro:154:30
```

### Describe the expected behaviour

```shell
Getting the gesture recogniser task accelerated by GPU, like in this example after selecting GPU Delegate:

https://mediapipe-studio.webapps.google.com/demo/gesture_recognizer
```

### Standalone code/steps you may have used to try to get what you need

```shell
https://github.com/weizenyang/ADSLWebInterface/blob/main/ADSLInterface/src/pages/index.astro
```

### Other info / Complete Logs

Any help will be highly appreciated. Thank you!

**Comment by ayushgdev** [comment]
Hello @weizenyang 
This is working fine. The link of ADSLWebInterface repo provided is not working. So please refer the following code for running an example with GPU delegate:

```html
<!DOCTYPE html>
<html>
<head></head>
  <body>
    <video width="1920" height="1080"  autoplay id="video">
    </video>
    <canvas id="canvas" ></canvas>
    <script type="module" src="../dist/gesture_recognition.js"></script>
  </body>
</html>
```

```javascript
// import { FilesetResolver, GestureRecognizer } from "@mediapipe/tasks-vision";
import {
    GestureRecognizer,
    FilesetResolver,
    DrawingUtils
} from "https://cdn.skypack.dev/@mediapipe/tasks-vision@0.1.0-alpha-11";

let gestureRecognizer = null
let canvas = document.getElementById('canvas')
// Create task for image file processing:
const createGestureRecognizer = async () => {
    const vision = await FilesetResolver.forVisionTasks(
        "https://cdn.jsdelivr.net/npm/@mediapipe/tasks-vision@0.1.0-alpha-11/wasm"
    );
    gestureRecognizer = await GestureRecognizer.createFromOptions(vision, {
        baseOptions: {
            delegate: 'GPU',
            modelAssetPath:
                "https://storage.googleapis.com/mediapipe-tasks/gesture_recognizer/gesture_recognizer.task"
        },
        runningMode: 'video',
        canvas: canvas
    });
    // getUsermedia parameters.
    const constraints = {
        video: true
    };

    // Activate the webcam stream.
    navigator.mediaDevices.getUserMedia(constraints).then(function (stream) {
        video.srcObject = stream;
        video.addEventListener("loadeddata", predictWebcam);
    });
}
function predictWebcam() {
    let nowInMs = Date.now();
    const results = gestureRecognizer.recognizeForVideo(video, nowInMs);

    console.log(results)
    window.requestAnimationFrame(predictWebcam);
}
createGestureRecognizer();

```

**Comment by eldog** [comment]
I encountered this problem when using the HandLandmarker when I was using a canvas which I had taken a context 2d from (e.g. I was calling `canvas.getContext('2d')`). Using a dedicated canvas solved it - I guess it's masking an internal error where you can't have both a WebGL and 2D context for a canvas (it returns null after the first).

**Comment by tyrmullen** [comment]
Correct; all our GPU processing is done with WebGL2, so any canvas passed to any Solutions API needs to either be uninitialized or have a WebGL2 context (since as you noted HTMLCanvasElement objects only ever initialize once, after which any `getContext` call requesting a different type will return `null`).  FWIW, uninitialized is generally ideal if it works for your use case, so that MediaPipe can initialize it fully and properly internally.

The crashing error here is the `emscripten_webgl_create_context` failing message, since that's how we initialize our WebGL context (Emscripten lets us make these web calls from C++ code), so the internal error isn't exactly masked, but it's not as user-friendly as we might like.

Actually, I'm not sure that we necessarily need to have the user make a canvas for GPU processing-- I'd expect that would get created automatically internally-- so one potential future improvement might be to allow GPU processing without requiring that the user specifies a canvas manually, in which case your very first attempt would succeed.

**Comment by weizenyang** [comment]
Hi all,

I finally came around to update on this project.

I've taken eldog's advice and initialized an empty canvas. Now it seems like the null error is gone, but now the error states like the Hand Gesture Recognizer only contains CPU ops. (Full error in image below). But from the Mediapipe example, this shouldn't be the case I think?

![image](https://github.com/google/mediapipe/assets/43901693/02676e32-cfc2-467b-85de-1b35138a6a8e)

I trained the model using my own weighting following this guide. I've also tested the model on Mediapipe's demo page and it seems to be utilizing the GPU for inference on the site.

One thing about my project is that I was also using @mediapipe/hands instead of the vision_tasks/HandLandmarker to display the hand landmarks. Maybe that might cause it? I chose the former as the performance seemed to be better on 2 Codepen projects I've tested (but in hindsight, maybe the vision_tasks/HandLandmarker Codepen was not coded to take advantage of GPU inference?? Although it seems to be fixed now).

@mediapipe/hands
https://codepen.io/mediapipe/pen/RwGWYJw

@mediapipe/HandLandmarker
https://codepen.io/mediapipe-preview/pen/gOKBGPN

I'll fork the project and try to use vision_task instead of 2 different implementations
from 2 different repos.
